tools/codegen/gen.py

```diff
@@ -1275,16 +1300,46 @@ def gen_source_files(
     for dispatch_key in dispatch_keys:
         fm = cuda_fm if is_cuda_dispatch_key(dispatch_key) else cpu_fm
 
+        if per_operator_headers:
+            def operator_headers() -> List[str]:
+                headers = []
+                for fn in native_functions:
+                    is_registered = backend_index.has_kernel(fn) or (
+                        fn.structured and dispatch_key in
+                        (DispatchKey.Meta, DispatchKey.CompositeExplicitAutograd))
+                    if not is_registered:
+                        continue
+
+                    headers.append(f"#include <ATen/ops/{fn.root_name}_native.h>")
+                    if dispatch_key == DispatchKey.CompositeExplicitAutograd:
+                        headers.append(f"#include <ATen/ops/{fn.root_name}.h>")
+                    if dispatch_key in functions_keys:
+                        headers.append(
+                            f"#include <ATen/ops/{fn.root_name}_{dispatch_namespace}_dispatch.h>")
+
+                return sorted(set(headers))
```

Comment:
Is there any particular reason you coalesced multiple overloads together into a single header by root_name? Besides aesthetics, it seems strictly worse, since there isn't any intrinsic relationship between various overloads and they can be profitably split into separate headers. Would probably simplify the grouping logic elsewhere as well.

Reply:
This might be good to discuss in pytorch/rfcs#39.  The main reason for me is I don't think a programmer should need to know the name of the overload they're calling, which you would in order to include it. Of course this is a trade off for more files potentially being rebuilt when an overload is changed but I am doubtful this is a significant issue e.g. most overloads are implemented in the same file so would get rebuilt together anyway.

I suppose there could be a hierarchy of headers where we start with one per overload and then group headers by `root_name` then into the top level header like  `Functions.h`. However, that would be a massive jump in file count which might lead to problems of their own.

Reply:
Well, I discovered one good reason to coalesce: it makes the logic here simpler for grouped native functions (patch coming soon)